File: 2019/R/d01.R
part1 <- function(mass) {
  floor(mass/3) - 2
}

part2 <- function(mass) {
  f <- floor(mass/3) - 2
  if (f > 0) f + part2(f)
  else 0
}

input <- as.integer(readLines("../inputs/input_1.txt"))
c(sum(part1(input)),
  sum(unlist(lapply(input, part2))))

File: 2015/python/d01.py
import numpy as np

def read(filename):
    with open(filename, 'r') as file:
         return(file.read().replace('\n', ''))

def convert(str):
    return([(41 - ord(x)) * 2 - 1 for x in str])

def part1(data):
    return sum(data)

def part2(data):
    return np.where(np.cumsum(data) < 0)[0][0] + 1

def tests():
    assert(part1(convert('(())')) == 0)
    assert(part1(convert('()()')) == 0)
    assert(part1(convert('(((')) == 3)
    assert(part1(convert('(()(()(')) == 3)
    assert(part1(convert('))(((((')) == 3)
    assert(part1(convert('())')) == -1)
    assert(part1(convert('))(')) == -1)
    assert(part1(convert(')))')) == -3)
    assert(part1(convert(')())())')) == -3)
    assert(part2(convert(')')) == 1)
    assert(part2(convert('()())')) == 5)

data = convert(read('../inputs/input_1.txt'))
print(part1(data))
print(part2(data))
